### PR TITLE
Support fill_value for ImageProjectiveTransform

### PR DESCRIPTION
**Pull request by WindQAQ**
As per https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/layers/preprocessing/image_preprocessing.py#L908.

cc @tanzhenyu

### CODE REVIEW

**Review by tanzhenyu — DISMISSED** [review]
Awesome! Thank you for the contribution.

**Comment by tanzhenyu** [comment]
@WindQAQ If you'd like to contribute more, I'd encourage you to implement the TPU kernel for this as well, this will give us the full spectrum

**Comment by WindQAQ** [comment]
Hi @tanzhenyu, I'll try it. If I understand correctly, I can add a XLA op in https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/tf2xla/kernels/image_ops.cc, right? Thank you for the review :-)

**Comment by tanzhenyu** [comment]
> Hi @tanzhenyu, I'll try it. If I understand correctly, I can add a XLA op in https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/tf2xla/kernels/image_ops.cc, right? Thank you for the review :-)

Yes that is correct

**Comment by WindQAQ** [comment]
> > Hi @tanzhenyu, I'll try it. If I understand correctly, I can add a XLA op in https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/tf2xla/kernels/image_ops.cc, right? Thank you for the review :-)
> 
> Yes that is correct

Got it, will do it in next PR.

**Comment by WindQAQ** [comment]
- Fix typo
- Add documentation
- Update golden API

**Review by tanzhenyu — DISMISSED** [review]

**Review by karmel — CHANGES_REQUESTED** [review]
for @tensorflow/api-owners 

This op is differentiable, so we need to make a new kernel for this (V3) that takes fill_value as an input rather than an attr, so that we can differentiate with regard to fill_value. With the new kernel, you will have to update the gradient registration as well. (This will have to have a forward compatibility guard as well: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/compat/compat.py @tanzhenyu can help with that as necessary.)

**Review by alextp — COMMENTED** [review]
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/image_ops.py#L242 also needs to be modified to register a gradient wrt the fill value for the v3 op

**Comment by WindQAQ** [comment]
> https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/image_ops.py#L242 also needs to be modified to register a gradient wrt the fill value for the v3 op

Thanks for the reminder! I am aware of it, will do it real quick :-)

**Review by alextp — DISMISSED** [review]

**Comment by WindQAQ** [comment]
@alextp can you approve it again? Just update some machine generated files. Thanks!

**Review by alextp — DISMISSED** [review]

**Review by alextp — DISMISSED** [review]

**Comment by WindQAQ** [comment]
Hi @alextp, is there any update on this? I've prepared an XLA kernel for `ImageProjectiveTransform` as per @tanzhenyu's comment and wanna register on V3 directly. Thank you!

**Comment by alextp** [comment]
@gbaned what's the status of merging this internally?

**Comment by gbaned** [comment]
> @gbaned what's the status of merging this internally?

@alextp This is waiting for internal merge. 
CC @tanzhenyu  Any update on this? Please. Thanks!

**Comment by tanzhenyu** [comment]
> @gbaned what's the status of merging this internally?

I think per our discussion, there is an issue with `pywarp_gradient_exclusion.cc` -- this seems a typo to me so it's messing up with the correct one, which is `pywrap_gradient_exclusions.cc`

**Comment by WindQAQ** [comment]
> > @gbaned what's the status of merging this internally?
> 
> I think per our discussion, there is an issue with `pywarp_gradient_exclusion.cc` -- this seems a typo to me so it's messing up with the correct one, which is `pywrap_gradient_exclusions.cc`

Umm, that's weird, seems that it's a machine generated file. I can delete it anyway.

**Comment by tanzhenyu** [comment]
> > > @gbaned what's the status of merging this internally?
> > 
> > 
> > I think per our discussion, there is an issue with `pywarp_gradient_exclusion.cc` -- this seems a typo to me so it's messing up with the correct one, which is `pywrap_gradient_exclusions.cc`
> 
> Umm, that's weird, seems that it's a machine generated file. I can delete it anyway.

What kind of cli you run to generate that file?

**Comment by WindQAQ** [comment]
> > > > @gbaned what's the status of merging this internally?
> > > 
> > > 
> > > I think per our discussion, there is an issue with `pywarp_gradient_exclusion.cc` -- this seems a typo to me so it's messing up with the correct one, which is `pywrap_gradient_exclusions.cc`
> > 
> > 
> > Umm, that's weird, seems that it's a machine generated file. I can delete it anyway.
> 
> What kind of cli you run to generate that file?

I forget about it. I just saw `Ubuntu`

> > > > @gbaned what's the status of merging this internally?
> > > 
> > > 
> > > I think per our discussion, there is an issue with `pywarp_gradient_exclusion.cc` -- this seems a typo to me so it's messing up with the correct one, which is `pywrap_gradient_exclusions.cc`
> > 
> > 
> > Umm, that's weird, seems that it's a machine generated file. I can delete it anyway.
> 
> What kind of cli you run to generate that file?

I forget about it. I remember I saw the CI emitted some errors and I followed the instructions on it. It might be something wrong at my end.

**Comment by WindQAQ** [comment]
CI still passes after deleting that file. Sorry that it's probably my fault.

@tanzhenyu and @alextp Can you review this again when time allows? Thank you!

**Comment by tanzhenyu** [comment]
Would you mind checking-in (or at least help) with this as well? https://github.com/tensorflow/tensorflow/pull/42077

**Comment by WindQAQ** [comment]
> Would you mind checking-in (or at least help) with this as well? #42077

I can take over it in this PR too if the author does mind. Let me ask him first.

**Comment by WindQAQ** [comment]
Hi @tanzhenyu, could you review it again when time allows? I also fix #42077.

File: tensorflow/python/keras/layers/preprocessing/image_preprocessing.py
Comment (tanzhenyu) [inline tensorflow/python/keras/layers/preprocessing/image_preprocessing.py:470]: Put `fill_value` at the end

**Review by tanzhenyu — COMMENTED** [review]
Thanks for the PR!
Does it resolve the dashed line problem?

**Comment by Smankusors** [comment]
> Thanks for the PR!
> Does it resolve the dashed line problem?

yeah according to https://github.com/tensorflow/tensorflow/pull/42077#issuecomment-670958925

to be honest I haven't tried his code though... but it seems his fix is better than mine

**Comment by tanzhenyu** [comment]
> > Thanks for the PR!
> > Does it resolve the dashed line problem?
> 
> yeah according to [#42077 (comment)](https://github.com/tensorflow/tensorflow/pull/42077#issuecomment-670958925)
> 
> to be honest I haven't tried his code though... but it seems his fix is better than mine

Great. I will approve once @WindQAQ moved `fill_value` to the end of arg list.

File: tensorflow/python/keras/layers/preprocessing/image_preprocessing.py
Comment (WindQAQ) [reply tensorflow/python/keras/layers/preprocessing/image_preprocessing.py:470]: do you mean also change the input signature for `__init__`?

```python
  def __init__(self,
               height_factor,
               width_factor,
               fill_mode='reflect',
               interpolation='bilinear',
               seed=None,
               name=None,
               fill_value=0.0,
               **kwargs):
```

**Review by tanzhenyu — DISMISSED** [review]

**Comment by gbaned** [comment]
@WindQAQ  Can you please resolve conflicts? Thanks!

File: tensorflow/core/ops/ops.pbtxt
Comment (tanzhenyu) [inline tensorflow/core/ops/ops.pbtxt:19126]: IIUC, what @alextp commented is that this file should not be modified.

Comment (WindQAQ) [reply tensorflow/core/ops/ops.pbtxt]: Solved.

**Review by tanzhenyu — APPROVED** [review]
@alextp Sorry for the back and forth, would you mind re-LGTM this?

**Comment by Smankusors** [comment]
Sorry if I'm asking this.... but.... I'm wondering when this will get merged?

Thanks

**Comment by tanzhenyu** [comment]
> Sorry if I'm asking this.... but.... I'm wondering when this will get merged?
> 
> Thanks

My guess would be today

**Comment by Smankusors** [comment]
> > Sorry if I'm asking this.... but.... I'm wondering when this will get merged?
> > Thanks
> 
> My guess would be today

it seems.... it takes longer than expected?

**Pull request merged**
